File: solutions/python/yacht/1/yacht.py
"""A function for calculating the score in a game of Yacht."""

# Score categories.
# Change the values as you see fit.
YACHT = "yacht"
ONES = 1
TWOS = 2
THREES = 3
FOURS = 4
FIVES = 5
SIXES = 6
FULL_HOUSE = "full house"
FOUR_OF_A_KIND = "four of a kind"
LITTLE_STRAIGHT = "little straight"
BIG_STRAIGHT = "big straight"
CHOICE = "choice"


def score(dice, category):
    match category:
        case "yacht":
            if len(set(dice)) == 1:
                return 50
        case "choice":
            return sum(dice)
        case "big straight":
            if set(dice) == {2,3,4,5,6}:
                return 30
        case "little straight":
            if set(dice) == {1,2,3,4,5}:
                return 30
        case "four of a kind":
            dice.sort()
            if len(set(dice[0:4])) == 1 or len(set(dice[1:5])) == 1:
                return 4 * dice[2]
        case "full house":
            dice.sort()
            if ((len(set(dice[0:3])) == 1 and len(set(dice[3:5])) == 1) or (len(set(dice[0:2])) == 1 and len(set(dice[2:5])) == 1)) and dice[1] != dice[3]:
                return 2 * dice[0] + dice[2] + 2 * dice[4]
        case number:
            return number * dice.count(number)
    return 0

File: solutions/fsharp/yacht/1/Yacht.fs
module Yacht

type Category = 
    | Ones
    | Twos
    | Threes
    | Fours
    | Fives
    | Sixes
    | FullHouse
    | FourOfAKind
    | LittleStraight
    | BigStraight
    | Choice
    | Yacht

type Die =
    | One = 1
    | Two = 2
    | Three = 3
    | Four = 4
    | Five = 5
    | Six = 6

let score category dice = 
    match category with
        | Category.Ones -> dice |> List.filter (fun x -> x = Die.One) |> List.fold (fun s c -> s + 1) 0
        | Category.Twos -> dice |> List.filter (fun x -> x = Die.Two) |> List.fold (fun s c -> s + 2) 0
        | Category.Threes -> dice |> List.filter (fun x -> x = Die.Three) |> List.fold (fun s c -> s + 3) 0
        | Category.Fours -> dice |> List.filter (fun x -> x = Die.Four) |> List.fold (fun s c -> s + 4) 0
        | Category.Fives -> dice |> List.filter (fun x -> x = Die.Five) |> List.fold (fun s c -> s + 5) 0
        | Category.Sixes -> dice |> List.filter (fun x -> x = Die.Six) |> List.fold (fun s c -> s + 6) 0
        | Category.FullHouse ->
            let groups = List.groupBy (fun x -> x) dice
            match groups with
                | (v1, [c1; c2; c3]) :: (v2, [c4; c5]) :: rest -> List.fold (fun s c -> s + int c) 0 dice
                | (v1, [c1; c2]) :: (v2, [c3; c4; c5]) :: rest -> List.fold (fun s c -> s + int c) 0 dice
                | _ -> 0
        | Category.FourOfAKind -> 
            let sum = List.filter (fun x -> x = Die.One) dice |> List.length
            if sum >= 4 then
                4
            else
                let sum = List.filter (fun x -> x = Die.Two) dice |> List.length
                if sum >= 4 then
                    8
                else
                    let sum = List.filter (fun x -> x = Die.Three) dice |> List.length
                    if sum >= 4 then
                        12
                    else
                        let sum = List.filter (fun x -> x = Die.Four) dice |> List.length
                        if sum >= 4 then
                            16
                        else
                            let sum = List.filter (fun x -> x = Die.Five) dice |> List.length
                            if sum >= 4 then
                                20
                            else
                                let sum = List.filter (fun x -> x = Die.Six) dice |> List.length
                                if sum >= 4 then
                                    24
                                else
                                    0
        | Category.LittleStraight ->
            if (List.contains Die.One dice) && (List.contains Die.Two dice) && (List.contains Die.Three dice) && (List.contains Die.Four dice) && (List.contains Die.Five dice) then
                30
            else
                0
        | Category.BigStraight ->
            if (List.contains Die.Six dice) && (List.contains Die.Two dice) && (List.contains Die.Three dice) && (List.contains Die.Four dice) && (List.contains Die.Five dice) then
                30
            else
                0
        | Category.Choice -> dice |> List.fold (fun s c -> s + int c) 0
        | Category.Yacht ->
            if List.forall (fun x -> x = Die.One) dice || List.forall (fun x -> x = Die.Two) dice || List.forall (fun x -> x = Die.Three) dice || List.forall (fun x -> x = Die.Four) dice || List.forall (fun x -> x = Die.Five) dice || List.forall (fun x -> x = Die.Six) dice then
                50
            else
                0